src/performanceEventBody.js

```diff
@@ -0,0 +1,31 @@
+import {EventBody} from './eventBody';
+
+function ensureArray(value) {
+	if (!Array.isArray(value)) {
+		return [value];
+	}
+	return value;
+}
+
+export class PerformanceEventBody extends EventBody {
+	constructor() {
+		super(...arguments);
+		super.setTypeGuid('02a00ca0-e67d-4a6c-908b-ee80b8a61eec');
+	}
+
```

Comment:
I wonder how closely we'd like to model this after [`PerformanceEntry`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceEntry)? All the different types of entries (mark, measure, paint, navigation, resource and frame) all inherit from it, and have a `entryType`, `name`, `startTime`, and `duration`.

Reply:
Additionally, it might be nice to be able to group entries by category/product? e.g. `my-courses`, `activity-feed`, etc. Or is that what the eventTypeGuid is?

Reply:
Regarding grouping by category/product, I believe the SourceId in the event envelope meets that need.
We've adopted the convention of setting SourceId equal to something like:
`https://schema.brightspace.com/ns/activityfeed`
We take the value after `ns` and use that to select the associated delivery stream which uses that value as the delivery stream prefix so that all data written to S3 is written to a folder matching that name. Then we can use that as the primary partition for athena queries.

Reply:
Regarding the reference to PerformanceEntry.
We are actually just logging the performance entries right now, but I didn't really check to see if different entries might have different properties. I have not restricted what can go in the UserTiming property.

Here is a full example of one of our UI perf events:

```
{
   "EventType":"https://schema.brightspace.com/ns/common/ui-performance-event.json",
   "SourceId":"https://schema.brightspace.com/ns/activityfeed",
   "Date":"2018-02-07T21:10:59.427Z",
   "EventBody":{
      "EventTypeGuid":"02a00ca0-e67d-4a6c-908b-ee80b8a61eec",
      "Action":"https://schema.brightspace.com/ns/common#PageLoad",
      "TenantUrl":"https://alafqa.devlms.brightspace.com/",
      "Context":{
         "Id":"123061",
         "Type":"Course Offering"
      },
      "Object":{
         "Id":"https://qa.activityfeed.us-west-2.brightspace.com/api/v1/d2l:orgUnit:123061/article/",
         "Type":"https://www.w3.org/ns/activitystreams#OrderCollectionPage"
      },
      "Custom":[
         {
            "name":"https://schema.brightspace.com/ns/common#PageLocation",
            "value":"https://qa.activityfeed.us-west-2.brightspace.com/"
         }
      ],
      "UserTiming":[
         {
            "name":"d2l.af.page.display",
            "entryType":"measure",
            "startTime":2,
            "duration":372.599999981001
         },
         {
            "name":"d2l.af.page.domInteractive",
            "entryType":"measure",
            "startTime":2,
            "duration":327
         },
         {
            "name":"d2l.af.api.profileFetchStart",
            "entryType":"measure",
            "startTime":2,
            "duration":544.0000000020955
         },
         {
            "name":"d2l.af.api.activitiesFetchStart",
            "entryType":"measure",
            "startTime":2,
            "duration":544.6000000014901
         },
         {
            "name":"d2l.af.page.rendered",
            "entryType":"measure",
            "startTime":2,
            "duration":1252.3999999761581
         },
         {
            "name":"d2l.af.api.profile",
            "entryType":"measure",
            "startTime":546.0000000020955,
            "duration":141.89999998779967
         },
         {
            "name":"d2l.af.api.token",
            "entryType":"measure",
            "startTime":546.4999999967404,
            "duration":112.69999999785796
         },
         {
            "name":"d2l.af.api.activities",
            "entryType":"measure",
            "startTime":546.6000000014901,
            "duration":241.99999999837019
         },
         {
            "name":"d2l.af.page.renderActivities",
            "entryType":"measure",
            "startTime":788.5999999998603,
            "duration":465.79999997629784
         }
      ],
      "Timestamp":"/Date 1518037859.982/",
      "Actor":{
         "Id":"30220",
         "Type":"User"
      },
      "Browser":{
         "Name":"Chrome",
         "MajorVersion":"64",
         "Minor":"0",
         "OS":"Windows",
         "OSVersion":"10.0",
         "Device":"Other"
      }
   },
   "name":"https://schema.brightspace.com/ns/common#PageLoad",
   "ts":1518037859,
   "Version":"0.1",
   "EventId":"657ef620-0c4b-11e8-9bb1-5bc92495dc8e",
   "Timestamp":"/Date 1518037859.427/",
   "TenantId":"b6427b6e-eaa7-4680-b168-1a438d739eda"
}
```